pkg/common/kafka/producer_test.go

```diff
@@ -0,0 +1,157 @@
+)
+
+var _ = Describe("Kafka Producer Test", func() {
+	var mockTopicService *mock_kafkacommon.MockTopicServiceInterface
+	var mockKafkaConfigService *mock_kafkacommon.MockKafkaConfigMapServiceInterface
+	var service *kafkacommon.ProducerService
+	var ctrl *gomock.Controller
+
+	BeforeEach(func() {
+		ctrl = gomock.NewController(GinkgoT())
+		mockTopicService = mock_kafkacommon.NewMockTopicServiceInterface(ctrl)
+		mockKafkaConfigService = mock_kafkacommon.NewMockKafkaConfigMapServiceInterface(ctrl)
+		service = &kafkacommon.ProducerService{
+			Topic:          mockTopicService,
+			KafkaConfigMap: mockKafkaConfigService,
+		}
+		config.Init()
+		authType := clowder.BrokerConfigAuthtype("Auth")
+		dummyString := gomock.Any().String()
+		mech := "PLAIN"
+		proto := "SASL_SSL"
+		port := 80
+		brokerConfig := clowder.BrokerConfig{
+			Authtype: &authType,
+			Cacert:   &dummyString,
+			Hostname: "192.168.1.7",
+			Port:     &port,
+			Sasl: &clowder.KafkaSASLConfig{
+				SaslMechanism:    &mech,
+				SecurityProtocol: &proto,
+				Username:         &dummyString,
+				Password:         &dummyString,
+			},
+		}
+		brokerSlice := []clowder.BrokerConfig{brokerConfig}
+		config.Get().KafkaBrokers = brokerSlice
+
+	})
+	AfterEach(func() {
+		ctrl.Finish()
+		service.UnsetProducer()
+	})
+	Describe("Get Producer instance", func() {
+		When("One dosent exist yet", func() {
+			Context("One dosent exist yet", func() {
+				It("should be ok", func() {
+					kafkaConfigMap := kafka.ConfigMap{
+						"bootstrap.servers": "192.168.1.3:80",
+					}
+					mockKafkaConfigService.EXPECT().GetKafkaProducerConfigMap().Return(kafkaConfigMap)
+					p := service.GetProducerInstance()
+					Expect(p).ToNot(BeNil())
+
+				})
+			})
+			Context("Producer is a singleton", func() {
+				It("should be the same producer", func() {
+					kafkaConfigMap := kafka.ConfigMap{
+						"bootstrap.servers": "192.168.1.1:80",
+					}
+					mockKafkaConfigService.EXPECT().GetKafkaProducerConfigMap().Return(kafkaConfigMap)
+					p1 := service.GetProducerInstance()
+					Expect(p1).ToNot(BeNil())
+
+					p2 := service.GetProducerInstance()
+					Expect(p2).ToNot(BeNil())
+					Expect(p1).To(Equal(p2))
+
+				})
+			})
+		})
+	})
+	Describe("Get Producer instance fails", func() {
+		When("One dosent exist yet two", func() {
+			Context("Kafka Config errors", func() {
+				It("Producer should be nil", func() {
+					kafkaConfigMap2 := kafka.ConfigMap{
+						"bootstrap.servers": "192.168.1.2:80",
+						"sasl.mechanisms":   gomock.Any(),
+						"security.protocol": gomock.Any(),
+					}
+					mockKafkaConfigService.EXPECT().GetKafkaProducerConfigMap().Return(kafkaConfigMap2)
+					p1 := service.GetProducerInstance()
+					Expect(p1).To(BeNil())
+				})
+			})
+		})
+	})
+	Describe("Produce Event", func() {
+		When("OK", func() {
+			Context("ok", func() {
+				It("should be ok", func() {
+					topic := "Test Topic"
+					mockTopicService.EXPECT().GetTopic(gomock.Any()).Return(topic, nil)
```

Comment:
the `GetTopic` parameter is `string`, I think would make sense to use the `topic` as expect value

Reply:
I dont understand what you mean by "expect value"

`topic` is the value I wish my mock to return, I dont care what values are passed to my mock, just return `topic` when called.

Reply:
my suggestion was just to use it like it below because if somehow you sent a different  value than the topic it will fail
so you are guaranteed that value will be the topic, and also mocking it
`
```suggestion
					mockTopicService.EXPECT().GetTopic(topic).Return(topic, nil)
```